freezer/src/freezer_files.rs

```diff
@@ -0,0 +1,511 @@
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn get_bounds(&self, item: u64) -> Result<Option<(u64, u64, FileId)>, IoError> {
+        let mut buffer = [0; INDEX_ENTRY_SIZE as usize];
+        let mut index = &self.index;
+        if let Err(e) = index.seek(SeekFrom::Start(item * INDEX_ENTRY_SIZE)) {
+            ckb_logger::trace!("Freezer get_bounds seek {} {}", item * INDEX_ENTRY_SIZE, e);
+            return Ok(None);
+        }
+
+        if let Err(e) = index.read_exact(&mut buffer) {
+            ckb_logger::trace!("Freezer get_bounds read_exact {}", e);
+            return Ok(None);
+        }
+        let end_index = IndexEntry::decode(&buffer)?;
+        if item == 1 {
+            return Ok(Some((0, end_index.offset, end_index.file_id)));
+        }
+
+        if let Err(e) = index.seek(SeekFrom::Start((item - 1) * INDEX_ENTRY_SIZE)) {
+            ckb_logger::trace!(
+                "Freezer get_bounds seek {} {}",
+                (item - 1) * INDEX_ENTRY_SIZE,
+                e
+            );
+            return Ok(None);
+        }
+        if let Err(e) = index.read_exact(&mut buffer) {
+            ckb_logger::trace!("Freezer get_bounds read_exact {}", e);
+            return Ok(None);
+        }
+        let start_index = IndexEntry::decode(&buffer)?;
+        if start_index.file_id != end_index.file_id {
+            return Ok(Some((0, end_index.offset, end_index.file_id)));
+        }
+
+        Ok(Some((
+            start_index.offset,
+            end_index.offset,
+            end_index.file_id,
+        )))
+    }
+
+    // keeping the the provided threshold number item and dropping the rest.
+    pub fn truncate(&mut self, item: u64) -> Result<(), IoError> {
+        // out of bound, this has no effect.
+        if item < 1 || ((item + 1) >= self.number()) {
+            return Ok(());
+        }
+        ckb_logger::trace!("Freezer truncate items {}", item);
+
+        let mut buffer = [0; INDEX_ENTRY_SIZE as usize];
+        // truncate the index
+        helper::truncate_file(&mut self.index, (item + 1) * INDEX_ENTRY_SIZE)?;
+        self.index.seek(SeekFrom::Start(item * INDEX_ENTRY_SIZE))?;
+        self.index.read_exact(&mut buffer)?;
+        let new_index = IndexEntry::decode(&buffer)?;
+
+        // truncate files
+        if new_index.file_id != self.head_id {
+            self.release(new_index.file_id);
+            let (new_head_file, offset) = self.open_append(new_index.file_id)?;
+
+            self.delete_after(new_index.file_id)?;
+
+            self.head_id = new_index.file_id;
+            self.head = Head::new(new_head_file, offset);
+        }
+        helper::truncate_file(&mut self.head.file, new_index.offset)?;
+        self.head.bytes = new_index.offset;
+        self.number.store(item + 1, Ordering::SeqCst);
+        Ok(())
+    }
+
+    pub fn preopen(&mut self) -> Result<(), IoError> {
+        self.release_after(0);
+
+        for id in self.tail_id..self.head_id {
```

Comment:
This opens all read-only files that have reached the max size, do you need to consider the lazy part to reduce the number of fd fetching?

Reply:
https://github.com/nervosnetwork/ckb/pull/2393